counsel.el

```diff
@@ -788,12 +788,26 @@ By default `counsel-bookmark' opens a dired buffer for directories."
       (put-text-property 0 (length key) 'face 'counsel-key-binding key)
       (format "%s (%s)" cmd key))))
 
-(defvar smex-initialized-p)
-(defvar smex-ido-cache)
-(declare-function smex-initialize "ext:smex")
-(declare-function smex-detect-new-commands "ext:smex")
-(declare-function smex-update "ext:smex")
-(declare-function smex-rank "ext:smex")
+
+(defvar smex-initialized-p nil)
+(defvar amx-initialized nil)
+
+(cond ((require 'smex nil 'noerror)
+       (defvar smex-ido-cache)
+       (declare-function smex-initialize "ext:smex")
+       (declare-function smex-detect-new-commands "ext:smex")
+       (declare-function smex-update "ext:smex")
+       (declare-function smex-rank "ext:smex")
+       (unless smex-initialized-p
+         (smex-initialize)))
+      ((require 'amx nil 'noerror)
+       (defvar amx-cache)
+       (declare-function amx-initialize "ext:amx")
+       (declare-function amx-detect-new-commands "ext:amx")
+       (declare-function amx-update "ext:amx")
+       (declare-function amx-rank "ext:amx")
+       (unless amx-initialized
+         (amx-initialize))))
```

Comment:
There are several issues with going this way:
1. It is better to load optional external packages on demand. The above will cause `smex.el` or `amx.el`, when available, to be loaded along with `counsel.el`.
2. The above will evaluate `smex-initialize`/`amx-initialize` when `counsel.el` is being loaded. This is a big no-no; libraries should strive to only introduce definitions and have as few side-effects as possible when being loaded, particularly if these are coming from third-party packages.
3. `(defvar VAR)` and `declare-function` are byte-compiler declarations, so they should be at the top-level.

Reply:
- About point 1 and 2:
I understand the effect of loading as I did in the code, but on the other hand do you think it is not too heavy to make a check trow a ``require'' and then another conditional in a so frequent call like  counsel-M-x? For almost any other function it shouldn't care, but M-x is like the centre of everything and ivy/counsel is supposed to be as light as possible as alternative to helm? 
- point 3
Some documentation about this, please?

Reply:
Is it possible, for example, to declare a local variable we set in the load process that we can check latter to avoid the excess of conditionals with complex calls (like require) in the M-x call?

Reply:
> do you think it is not too heavy to make a check trow a ``require'' and then another conditional in a so frequent call like counsel-M-x?

`require` isn't heavy at all; what's heavy is loading (i.e. opening and evaluating) the corresponding library, but this only happens once, after which `require` boils down to a negligible lookup in a list of symbols, specifically `(memq FEATURE features)`; see [`(elisp) Named Features`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Named-Features.html).

> For almost any other function it shouldn't care, but M-x is like the centre of everything

<kbd>M-x</kbd> is an interactive command, which means it is far from the centre of everything. It **only** gets called when the user invokes it. Compare this to hooks like `before-change-functions`, `after-change-functions`, `pre-command-hook`, `post-command-hook`, etc. which get called several times *every single time* something meaningful happens in Emacs, such as after every single key press.

What's more, calling `require` is completely negligible compared to filtering and formatting all the candidates to display in the minibuffer for `counsel-M-x`.

> * point 3
> Some documentation about this, please?

The documentation for `declare-function` is at [`(elisp) Declaring Functions`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Declaring-Functions.html). Its purpose is to inform the byte-compiler that calls to a particular function in the current file are kosher, even if they do not seem so at first. The byte-compiler is quite naive in this respect, so we always strive to make things as clear as possible to it. As such, conditional function declarations make little sense; you don't need to declare an external function before calling it, and declaring a function won't save you from a `void-function` error if the function is not actually defined.

The documentation for `defvar` is at [`(elisp) Defining Variables`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Defining-Variables.html). The placement of `(defvar VAR)`, however, matters much more that that of `declare-function`, as it marks the given symbol as special only within its lexical scope; see [`(elisp) Compiler Errors`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Compiler-Errors.html) and [`(elisp) Dynamic Binding Tips`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Dynamic-Binding-Tips.html). Again, there is little sense or function in hiding these declarations from the byte-compiler; besides, we want the `amx-*` and `smex-*` vars to be special globally.

In common usage, the only time `defvar` and `declare-function` don't appear at the literal top-level is when they are wrapped in `eval-when-compile`, so that they only take effect during byte-compilation and Elisp interpretation, not during byte-code evaluation.

> Is it possible, for example, to declare a local variable we set in the load process that we can check latter to avoid the excess of conditionals with complex calls (like require) in the M-x call?

As mentioned, `require` already does this for us. Loading `amx` or `smex` with `require` records the feature in the symbol list `features` so that the feature is only loaded once.

Reply:
See also [`(elisp) Coding Conventions`](https://www.gnu.org/software/emacs/manual/html_node/elisp/Coding-Conventions.html) re: loading libraries.